In [ ]:
import numpy
import nibabel as nib
import os
import matplotlib.pyplot as plt
# Importing our own functions
from tissue2mrprop.functions.volume import volume
from tissue2mrprop.functions.label import SegmentationLabel

# Example running the code
In this notebook will guide on how to use a segmentation file and obtain MR property volumes from it

In [10]:
# Inside the example folder we have already segmented files from CT
# dataset by Gatidis et al. https://doi.org/10.1038/s41597-022-01718-3
# Attribution to the data set link : https://creativecommons.org/licenses/by/4.0/
ct_wb_seg = nib.load("example/example_seg.nii.gz") # wb stands for whole body
# We can acquire the numpy data from the nifti
labels = ct_wb_seg.get_fdata()
# And we can quickly see the shape of our data
labels.shape

(512, 512, 828)

# MR Property Phantom Creation

In [11]:
# Here we call the Volume function
# This volume function works for whole body - Total Seg CT
new_vol = volume(ct_wb_seg)

In [9]:
# Given that we know that all the labels from Total Seg should be the same
# We can set all the labels automatically with a specific method
new_vol.create_segmentation_labels("TotalSegmentator","v2")

Label ID 256 not found.
Label ID 256 not found.
Label ID 289 not found.
Label ID 289 not found.


SegmentationLabelManager == Volume


In [5]:
dicc = new_vol.group_seg_labels("TotalSeg_CT","mod0")

In [6]:
new_vol.segmentation_labels

{0: <functions.label.SegmentationLabel at 0x28c21d005e0>,
 1: <functions.label.SegmentationLabel at 0x28c21d00e80>,
 2: <functions.label.SegmentationLabel at 0x28c21d01390>,
 3: <functions.label.SegmentationLabel at 0x28c21d018a0>,
 4: <functions.label.SegmentationLabel at 0x28c21d01db0>,
 5: <functions.label.SegmentationLabel at 0x28c21d022c0>,
 6: <functions.label.SegmentationLabel at 0x28c21d03d60>,
 7: <functions.label.SegmentationLabel at 0x28c21d03850>,
 8: <functions.label.SegmentationLabel at 0x28c21d03340>,
 9: <functions.label.SegmentationLabel at 0x28c21d02e30>,
 10: <functions.label.SegmentationLabel at 0x28c21d02920>,
 11: <functions.label.SegmentationLabel at 0x28c21cfe230>,
 12: <functions.label.SegmentationLabel at 0x28c21cfe740>,
 13: <functions.label.SegmentationLabel at 0x28c21cfec50>,
 14: <functions.label.SegmentationLabel at 0x28c21cff160>,
 15: <functions.label.SegmentationLabel at 0x28c21cff6d0>,
 16: <functions.label.SegmentationLabel at 0x28c21cffbe0>,
 17: <f

In [7]:
for key,value in new_vol.look_up.items():
            # Key is the number of ID and value is (name, sus)
    
    print(f"id {key}, label inf. {value}")

id 0, label inf. ('air', 0.35)
id 1, label inf. ('spleen', -9.05)
id 2, label inf. ('kidney', -9.05)
id 3, label inf. ('kidney', -9.05)
id 4, label inf. ('organ', -9.05)
id 5, label inf. ('liver', -9.05)
id 6, label inf. ('organ', -9.05)
id 7, label inf. ('organ', -9.05)
id 8, label inf. ('gland', -9.05)
id 9, label inf. ('gland', -9.05)
id 10, label inf. ('lung', 0.2)
id 11, label inf. ('lung', 0.2)
id 12, label inf. ('lung', 0.2)
id 13, label inf. ('lung', 0.2)
id 14, label inf. ('lung', 0.2)
id 15, label inf. ('esophagus', -9.05)
id 16, label inf. ('trachea', 0.2)
id 17, label inf. ('gland', -9.05)
id 18, label inf. ('organ', -9.05)
id 19, label inf. ('organ', -9.05)
id 20, label inf. ('organ', -9.05)
id 21, label inf. ('organ', -9.05)
id 22, label inf. ('organ', -9.05)
id 23, label inf. ('kidney', -9.05)
id 24, label inf. ('kidney', -9.05)
id 25, label inf. ('bone', -9)
id 26, label inf. ('bone', -9)
id 27, label inf. ('bone', -9)
id 28, label inf. ('bone', -9)
id 29, label inf. ('

In [8]:
for key,value in new_vol.look_up.items():
            # Key is the number of ID and value is (name, sus)
    
    print(f"id {key}, label inf. {value}")

id 0, label inf. ('air', 0.35)
id 1, label inf. ('spleen', -9.05)
id 2, label inf. ('kidney', -9.05)
id 3, label inf. ('kidney', -9.05)
id 4, label inf. ('organ', -9.05)
id 5, label inf. ('liver', -9.05)
id 6, label inf. ('organ', -9.05)
id 7, label inf. ('organ', -9.05)
id 8, label inf. ('gland', -9.05)
id 9, label inf. ('gland', -9.05)
id 10, label inf. ('lung', 0.2)
id 11, label inf. ('lung', 0.2)
id 12, label inf. ('lung', 0.2)
id 13, label inf. ('lung', 0.2)
id 14, label inf. ('lung', 0.2)
id 15, label inf. ('esophagus', -9.05)
id 16, label inf. ('trachea', 0.2)
id 17, label inf. ('gland', -9.05)
id 18, label inf. ('organ', -9.05)
id 19, label inf. ('organ', -9.05)
id 20, label inf. ('organ', -9.05)
id 21, label inf. ('organ', -9.05)
id 22, label inf. ('organ', -9.05)
id 23, label inf. ('kidney', -9.05)
id 24, label inf. ('kidney', -9.05)
id 25, label inf. ('bone', -9)
id 26, label inf. ('bone', -9)
id 27, label inf. ('bone', -9)
id 28, label inf. ('bone', -9)
id 29, label inf. ('

This Labels not found are pre-defined labels that work with final_sc_seg.nii.gz </br>
This file will be used later in the example of creating new labels :)

In [6]:
%%time
# Now we have all the labels from the whole body stablished
# By default T2star and Susceptibility values are complete 
# You can check the values used in the look-up table on the README file
sus_dist = new_vol.create_sus_dist()
# The output will be a Numpy array 

CPU times: total: 3.55 s
Wall time: 16.9 s


In [7]:
# A method is created to save the susceptibility from Numpy array to Nifti file
new_vol.save_sus_dist_nii()
# By default it will save it in the "output" folder

In a similar way we can create proton density and T2star volumes, acquire their numpy type data and then save them as nifti files

In [8]:
%%time
pd_vol = new_vol.create_pd_vol()
t2_star_vol = new_vol.create_t2_star_vol()

CPU times: total: 5.05 s
Wall time: 31.9 s


In [9]:
new_vol.save_pd_dist()
new_vol.save_t2star_dist()

Now you will be able to open any of this 3 volumes on a viewer of choice. </br>
As mentioned in the README file, if you use ITK snap you can open the segmentation labels.

# Label creation

This section will guide you through how to create new labels.
If you want, you can manually add a label in any desired location. For what you will only need a mask. </br>

This examples addresses a, as of June 2024, current limitation from Total Seg CT, that is that we cannot differentiate CSF, WM or GM from the Spinal Cord. </br>

Given that the labeling from anatomical image will have a specific value, when we add a new label we must add it with a value that is not included in the look up table. For example, for CT whole-body Total-Seg we must use a number higher than 117 to avoid conflicts. </br>


Given the nature of my example I used Spinal Cord Toolbox (SCT): https://spinalcordtoolbox.com/index.html by De Leener B, Levy S, Dupont SM, Fonov VS, Stikov N, Louis Collins D, Callot V, Cohen-Adad J. SCT: Spinal Cord Toolbox, an open-source software for processing spinal cord MRI data. Neuroimage 2017. </br> 
In order to add them we need to first register the CT image to the PAM50 space. For this we can use the Spinal Cord segmentation and you can manually create landmarks of spine C2 and C5. With this files ready you can automatically register to PAM50 space and procede with creating new labels. </br> 

**I.** The trick to add a new label to semgnetation is to first invert the values of the mask so that the ROI has value of 0 and everywhere else 1. </br> 
We then create a whole where we want the mask to be by multiplying it with our main labeled file.

In [ ]:
sct_maths -i new_label.nii.gz -o inv.nii.gz -sub 1
sct_maths -i inv.nii.gz -o inv.nii.gz -mul -1
sct_maths -i temp.nii.gz -o no_label.nii.gz -mul inv.nii.gz

**II.** We need to make sure that the new label will have a non-conflictng value depending on the range of the look-up table. </br>
*Note that we use temp as it is better to create a copy and edit the copy instead of the original.*

In [10]:
#Now the body will have zeroes where the label is going to go
# We need to set the value of the label to a number that 
# does not interfere with the look-up table
# For this example, any number higher than 117 would work 
sct_maths -i reslice_label.nii.gz -o new_label.nii.gz -mul 16
sct_maths -i new_label.nii.gz -o new_label.nii.gz -mul 16

We choose to multiply twice because there is a maximum value of 40 to multiply in SCT_maths. We can then create a new nifti file with value of 256 for the mask and 0 elsewhere. </br>
**III.** Then we just need to add the new label file to the file with a whole from the label we want to input.

In [ ]:
sct_maths -i no_label.nii.gz -o pro_labels.nii.gz -add new_label.nii.gz

## Important notes
- This procedure is not exclusive to SCT and can be perform with any tool of your preference!
- Once a new label is created, if you want to add more labels you will have to work on top of any previously made new volume with added labels, so creating a temporal copy to test first is recommended. </br>
- Also recommended, check the mask before multplication as I have deleted some files by multiplying inappropriate inversed masks. </br>
- In the "example" folder you will find an example of a succesfull implementation of 2 new masks to the CT_wb data that adds labels 289 and 256 to the whole body dataset.